### PR TITLE
8300208: Optimize Adler32 stub for AVX-512 targets.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Patch optimizes Adler32 stub for AVX512 target.

Main computation loop now uses zero extended lane widening load vector operation.

New sequence also honors AVX3Thresholds so that implementation uses existing AVX2 instruction sequence on relevant targets
if input size is smaller than threshold limit (default 4096).

Following are the result of an [existing JMH micro ](https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/java/util/TestAdler32.java)on various targets.

**System Configurations : Turbo frequency scaling is disabled, all the data is collected at fixed frequency of 2.8 GHz.
SUT1   : Intel® Xeon® Platinum 8480+ Processor (Sapphire Rapids)  56C 2S
SUT2   : Intel(R) Xeon(R) Platinum 8380 CPU (Icelake Server) 40C 2S
SUT3   : Intel(R) Xeon(R) Platinum 8280 CPU (Cascadelake Server) 28C 2S**

image

image

image

Please review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300208](https://bugs.openjdk.org/browse/JDK-8300208): Optimize Adler32 stub for AVX-512 targets.

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12045/head:pull/12045` \
`$ git checkout pull/12045`

Update a local copy of the PR: \
`$ git checkout pull/12045` \
`$ git pull https://git.openjdk.org/jdk pull/12045/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12045`

View PR using the GUI difftool: \
`$ git pr show -t 12045`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12045.diff">https://git.openjdk.org/jdk/pull/12045.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Looks good to me. Let me test it.

**Review by vnkozlov — APPROVED** [review]
My testing passed. You need second review.

File: test/micro/org/openjdk/bench/java/util/TestAdler32.java
Comment (turbanoff) [inline test/micro/org/openjdk/bench/java/util/TestAdler32.java:39]: nit:
```suggestion
    @Param({"64", "128", "256", "512", "1024", "2048", "5012", "8192", "16384", "32768", "65536"})
```

**Comment by jatin-bhateja** [comment]
Hi @sviswa7, can you kindly be second reviewer.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:147]: These two compares can be merged into one compare with larger of avx3_threshold() or 128.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:155]: This still processes 16 bytes worth of data in one loop iteration as the AVX2 loop. Have you given thoughts on processing double the size with AVX3?

**Comment by sviswa7** [comment]
Could you please also update the test/hotspot/jtreg/compiler/intrinsics/zip/TestAdler32.java to throw Exception on failure?

File: src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:153]: If we double the size of inputs processed within the loop from 16 to 32 then subsequent ADLER_SCALE will also change from 0-31, given that we cannot accommodate more than 16 integer in a 512 bit vector hence it will not be possible to scale up the loop block size.

**Comment by jatin-bhateja** [comment]
Hi @sviswa7 , your comments have been addressed.

**Comment by sviswa7** [comment]
@jatin-bhateja There are some windows build failures. Please take a look.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:153]: Could be done something like below:
[adler32.patch](https://github.com/openjdk/jdk/files/10521481/adler32.patch)

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp]: Yes, one level of unrolling will save loop jump back costs, I have incorporated this in the patch.

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
This needs to be tested before integration since stub code was changes since v00.

**Comment by jatin-bhateja** [comment]
> This needs to be tested before integration since stub code was changes since v00.

Hi @vnkozlov , please let me know once your test runs are through.

**Review by vnkozlov — APPROVED** [review]
My testing of latest version passed.

**Comment by vnkozlov** [comment]
@jatin-bhateja this changes seem cause regression https://bugs.openjdk.org/browse/JDK-8302358
Please, look.

**Pull request closed without merging**
